packager/packager.cc

```diff
@@ -438,8 +422,16 @@ bool CreateRemuxJobs(const StreamDescriptorList& stream_descriptors,
       if (hls_playlist_name.empty())
         hls_playlist_name = base::StringPrintf("stream_%d.m3u8", stream_number);
 
-      muxer_listener.reset(new HlsNotifyMuxerListener(hls_playlist_name, name,
-                                                      group_id, hls_notifier));
+      if(mpd_notifier && hls_notifier)
+          muxer_listener.reset(new HlsMpdNotifyMuxerListener(hls_playlist_name, name,
+                                                          group_id, hls_notifier, mpd_notifier));
+      else
+          muxer_listener.reset(new HlsNotifyMuxerListener(hls_playlist_name, name,
+                                                          group_id, hls_notifier));
+    } else if (mpd_notifier) {
+      std::unique_ptr<MpdNotifyMuxerListener> mpd_notify_muxer_listener(
+          new MpdNotifyMuxerListener(mpd_notifier));
+      muxer_listener = std::move(mpd_notify_muxer_listener);
     }
```

Comment:
Thanks for working on this. I'd prefer making code more generic. 

(1) We can have a CombinedMuxerListener, which contains one or more muxer listeners:

```C++
class CombinedMuxerListener : public class {
 public:
  explicit CombinedMuxerListener(std::list<std::unique_ptr<MuxerListener>>* muxer_listeners);

 private:
  CombinedMuxerListener(const CombinedMuxerListener&) = delete;
  CombinedMuxerListener& operator=(const CombinedMuxerListener&) = delete;

  std::list<std::unique_ptr<MuxerListener>> muxer_listeners_;
};

CombinedMuxerListener::CombinedMuxerListener(std::list<std::unique_ptr<MuxerListener>>* muxer_listeners) {
  DCHECK(muxer_listeners);
  muxer_listeners_.swap(*muxer_listeners);
}
```

For every function defined in MuxerListener interface, just forward to the members in the list, similar to what you did, e.g.

```C++
void CombinedMuxerListener::OnMediaStart(const MuxerOptions& muxer_options,
                                                                        const StreamInfo& stream_info,
                                                                        uint32_t time_scale,
                                                                        ContainerType container_type) override {
  for (auto& listener: muxer_listeners_) {
    listener->OnMediaStart(muxer_options, stream_info, time_scale, container_type);
  }
}
```

(2) The code can be changed like:

```C++
std::list<std::unique_ptr<MuxerListener>> muxer_listeners;

// For media info
muxer_listeners.emplace_back(new VodMediaInfoDumpMuxerListener(...));

// For MPD
muxer_listeners.emplace_back(new MpdNotifyMuxerListener(...));

// For HLS
muxer_listeners.emplace_back(new HlsNotifyMuxerListener(...));

// Then 
if (!muxer_listeners.empty()) {
  muxer->SetMuxerListener(std::unique_ptr<MuxerListener>(new CombinedMuxerListener(&muxer_listeners));
}
```